### PR TITLE
Add cards component

### PR DESCRIPTION
**Pull request by hannalaakso**
## What
The Find & View team created a local cards component for the homepage in [`frontend`](https://github.com/alphagov/frontend).

The component renders cards, each with a link and a description, as an unordered list element and visually presents it as a grid. It doesn't use the [GOV.UK Design System grid](https://design-system.service.gov.uk/styles/layout/#grid-system) but sets its own so that it can separately target desktop, tablet and mobile columns.

This PR:
- Ports the code over from  `frontend` 
- Updates markup to follow BEM conventions
- Adds print styles
- Adds automated template tests
- Allow the grid to be overridden to use two columns on desktop (this variant is currently used in the government activity sections of the homepage)
- Allows heading and subheading levels to be overridden
- Throws an error if any of the links are missing a `href` 
- Adds tracking attributes to the component card links
- Includes a heading element for the component to encourage best practise by having a heading for the landmark 
- Calculates rows with CSS grid and removes the IE11 polyfill: unlike the frontend repo, the new topic page templates can contain an unknown number of items so we're now using CSS grid to calculate rows and their heights, and falling back to a single column for IE11 -  [more information](https://github.com/alphagov/govuk_publishing_components/pull/2624/commits/0212604698d2076ab0445b12010d251bc3ce1268)

### MacOS Google Chrome
Screenshot 2022-03-02 at 19 20 30

### MacOS Firefox with custom colours
Screenshot 2022-03-02 at 19 21 31

### Windows 8 Internet Explorer 11
Screenshot 2022-03-02 at 19 37 44

### iOS 15 Safari

Screenshot 2022-03-02 at 19 35 59

## Why
The component is currently used in the list of topics and government activity sections on the homepage. 

We want to reuse it for Mainstream Browse pages so it needs to be made available to other apps via the gem.

## Things to note
- `.component-guide-preview` adds some padding which could make the cards component spacing seem slightly misaligned when comparing with the component on the homepage. I ended up temporarily removing the preview padding when I was checking the component spacing to get around this. 
- At the moment the `get_heading_level` helper is used to set the main heading level for the component. The ability to adjust the subheading level is manually configured in the template. It might be good to look at whether the helper should in fact have the ability to set more than one heading level for the same template, in which case the different heading levels could be linked to help ensure the correct heading level hierarchy. 

#### Windows	
- [x] Internet Explorer 11	compliant
- [x]  	Edge   97	compliant
- [x]  	Google Chrome (latest versions)	compliant
- [x]  	Mozilla Firefox (latest versions)	compliant

The chevron icon doesn't render correctly on IE8 but we've decided to leave it as is as it's still usable there.

#### macOS	
- [x] Safari 12 & 14	compliant
- [x]  	Google Chrome (latest versions)	compliant
- [x]  	Mozilla Firefox (latest versions)	compliant, including **when changing colours**

#### iOS	
- [x] Safari for iOS 15 	compliant
- [x]  	Google Chrome iOS 15 (latest versions)	compliant

#### Android	
- [x] Google Chrome (latest versions)	compliant
- [x]  	Samsung Internet (latest versions)	compliant

### Assistive tech to test with
- [x] JAWS (desktop screen reader)	18 or later	Chrome (latest version)
- [x] NVDA (desktop screen reader)	Latest	Firefox (latest version)
- [x] VoiceOver on iOS (mobile screen reader)	Latest	Safari (version 12 or later) 
- [x] TalkBack (mobile screen reader)	Latest	Chrome (latest version)
- [x] Windows Magnifier or Apple Zoom (screen magnifiers)	
- [ ] Dragon (speech recognition)	15 or later	Internet Explorer 11

## Other 
 - [x] Design review - see Trello card

### CODE REVIEW

File: app/views/govuk_publishing_components/components/docs/cards.yml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+  This component is currently experimental. If you are using it, please feed back any research findings to the GOV.UK Explore team.
+accessibility_criteria: |
+  The component must:
+
+  -  use the correct heading level hierarchy eg. if the component heading uses a`<h2>`, the subheadings must use a `<h3>`
+
+  The component should:
+
+  -  set the component heading to tell users what the list contains
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/cards.yml:18]: This acceptance criteria seems to imply that the heading is always required, but in the code it's an optional parameter. Is it always required?

File: app/views/govuk_publishing_components/components/docs/cards.yml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    data:
+      heading: Topics
+      heading_level: 3
+      sub_heading_level: 4
+      items:
+        - link:
+            text: Benefits
+            path: http://www.gov.uk
+          description: Includes eligibility, appeals, tax credits and Universal Credit
+  with_tracking:
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/cards.yml:125]: Usually we try to avoid examples that explicitly describe themselves as adding tracking, partly because that can raise privacy concerns but also because this doesn't appear to actually add tracking - the component would still need to be wrapped in a container with the `gem-track-click` module applied for tracking to occur. Maybe `with_link_data_attributes` instead?

File: app/assets/stylesheets/govuk_publishing_components/components/_cards.scss [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  padding: govuk-spacing(3) 0 govuk-spacing(6) 0;
+  position: relative;
+}
+
+.gem-c-cards__sub-heading {
+  margin: 0 govuk-spacing(6) govuk-spacing(2) 0;
+}
+
+.gem-c-cards__link {
+  &:after {
```
Comment (andysellick) [inline app/assets/stylesheets/govuk_publishing_components/components/_cards.scss:48]: I think the reason for this pseudo element is to make the entire area of the card clickable, is that correct? Is it worth a quick comment here to explain that?

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  sub_heading_level = sub_heading_level || 3
+  two_column_layout ||= false
+  local_assigns ||= nil
+  shared_helper = GovukPublishingComponents::Presenters::SharedHelper.new(local_assigns)
+%>
+<% if items.present? %>
+  <div class="gem-c-cards">
+    <% if heading %>
+      <%= content_tag(shared_helper.get_heading_level, class: "gem-c-cards__heading govuk-heading-m") do %>
+        <%= heading %>
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:13]: I included it because there was an ask from some of our stakeholders to just use the grid on a particular page without including a heading above it. The heading wasn't seen as necessary, maybe because a grid of items can to a sighted user seem like a self-contained unit. But semantically it's a list so including a heading which explains what the list contains could help users, especially screen reader ones if they are not able to glance over the grid to get a sense of its contents.

So basically the heading is inside the component to encourage best practise so that teams always consider including a heading as part of the component as it's:
- a part of the component and the documentation
- used in all the examples

But I'd be interested in hearing if there's a different convention in this repo to solve this.

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        end
+
+        data_attributes ||= nil
+        attributes = link[:tracking_attributes] || nil
+
+        if attributes && attributes[:track_category] && attributes[:track_action]
+          data_attributes = {
+            track_action: attributes[:track_action],
+            track_category: attributes[:track_category],
+            track_dimension_index: attributes[:track_dimension_index],
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:32]: Yes, this what our performance analyst specified and also what the select component in this repo does. 

There's some more detail in the commit message: https://github.com/alphagov/govuk_publishing_components/pull/2624/commits/66820768c57017eeb907f1b8d06ab6f5413c69c0

**Review by injms — CHANGES_REQUESTED** [review]
I don't think that this is the correct approach for a component - it's currently completely inflexible as it will only accept 16 items. Less than 16 items and there'll be a large gap - more than 16 and there will be overlapping cards.

This is a problem as the topics and subtopics pages where this is going to be used need a degree of flexibility as there will be a variable number of items that need to be placed.

I think that this should be just the (singular) card component without the grid layout. The frontend applications can then deal with the complications caused by the grid layout.

File: spec/components/cards_spec.rb [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select ".gem-c-cards__heading.govuk-heading-m", text: "Topics"
+  end
+
+  it "renders list heading using default heading level" do
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb:43]: Ideally unit tests should test a single logical outcome. So if the default heading level functionality breaks, having two isolated tests will assert that it is the broken default heading level that needs fixing whilst the heading classes will still be asserted to be correct. 

And if the requirement to use a default heading level here was to be removed in the future and this test was called `renders list heading using default heading level`, it would be easy for the developer to think that the test can be removed and overlook the fact that it is also used to assert the correct heading classes.

File: spec/components/cards_spec.rb [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select "h3.gem-c-cards__heading.govuk-heading-m", count: 1
+  end
+
+  it "renders list" do
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb:76]: I think again that this wouldn't follow the single outcome principle for unit tests (see https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811979265).

File: spec/components/cards_spec.rb [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select "ul.gem-c-cards__list.gem-c-cards__list--two-column-desktop", count: 1
+  end
+
+  it "renders a list item" do
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb:107]: I think again that this wouldn't follow the single outcome principle for unit tests (see https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811979265).

File: app/views/govuk_publishing_components/components/docs/cards.yml [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+          description: Tools and guidance for businesses
+        - link:
+            text: Business and self-employed
+            path: http://www.gov.uk
+          description: Tools and guidance for businesses
+        - link:
+            text: Business and self-employed
+            path: http://www.gov.uk
+          description: Tools and guidance for businesses
+  two_column_layout:
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml:70]: I think again that this wouldn't follow the single outcome principle for unit tests (see https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811979265).

File: spec/components/cards_spec.rb [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select ".govuk-link.gem-c-cards__link[data-track-action='some_action']", count: 1
+    assert_select ".govuk-link.gem-c-cards__link[data-track-category='homepageClicked']", count: 1
+    assert_select ".govuk-link.gem-c-cards__link[data-track-dimension-index='29']", count: 1
+  end
+
+  it "renders link text for data-track-dimension attribute" do
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb:222]: Yes, that's a good point, the dependency between the `data-track-dimension` and the link text is covered by the other test. I guess this test was to make that dependency even clearer if it breaks but I think `renders tracking attributes` is actually enough to cover and document it. 

I'll remove this test.

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:13]: I included it because there was an ask from some of our stakeholders to just use the grid on a particular page without including a heading above it. The heading wasn't seen as necessary, maybe (?) because a grid of items can seem like seem like a self-contained unit a sighted user. But semantically it's a list so including a heading which explains what the list contains could help users, especially screen reader users if they are not able to glance over the grid to get a sense of its contents.

So basically the heading is inside the component to encourage best practise so that teams always consider including a heading as part of the component as it's:
- a part of the component and the documentation
- used in all the examples

But I'd be interested in hearing what you think.

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 11]
```diff
@@ -0,0 +1,59 @@
+<%
+  heading ||= nil
+  items ||= nil
+  sub_heading_level = sub_heading_level || 3
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:4]: I mentioned this same issue in the PR description ('Things to note'). It seems like calculating the heading level could be a future improvement because the `get_heading_level` helper doesn't return an integer but a string so we'd need to do some splitting of strings. It seems like that the logic for that would be better either as a shared helper or maybe returned by `get_heading_level` somehow. 

Or let me know if I have in fact overlooked a way to easily get the integer from `get_heading_level`, I didn't too deeply into it.

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            track_label: link[:path],
+          }
+        end
+
+        link = capture do
+           link_to(link[:text], link[:path], {
+             class: "govuk-link gem-c-cards__link",
+             data: data_attributes,
+           })
+        end %>
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:43]: Could you check if this comment refers to the correct line? I think this particular tag is required here.

File: spec/components/cards_spec.rb [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select "h4.gem-c-cards__sub-heading.govuk-heading-s", text: "Benefits"
+  end
+
+  it "renders a link with href" do
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb:153]: I think again that this wouldn't follow the single outcome principle for unit tests (see https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811979265).

File: app/views/govuk_publishing_components/components/docs/cards.yml [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+  The component should:
+
+  -  set the component heading to tell users what the list contains
+shared_accessibility_criteria:
+  - link
+examples:
+  default:
+    data:
+      heading: Topics
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml:28]: This relates to the open comment on https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811984893.

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 1 above)
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml:22]: This relates to the open comment on https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811984893.

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 2 above)
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml:129]: We already use 'With tracking' on other component examples like https://components.publishing.service.gov.uk/component-guide/select/with_tracking. But happy to go with your suggestion, it seems sensible 👍

File: app/assets/stylesheets/govuk_publishing_components/components/_cards.scss (hunk 3 above)
Comment (hannalaakso) [reply app/assets/stylesheets/govuk_publishing_components/components/_cards.scss:56]: Yes definitely, I'll leave a comment.

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    <ul class="gem-c-cards__list<% if two_column_layout %> gem-c-cards__list--two-column-desktop<% end %>">
+      <% items.each do |item|
+
+        link = item[:link]
+
+        if !link[:path].present?
+          raise ArgumentError, "The cards component requires a href for all the links"
+        end
+
+        data_attributes ||= nil
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:25]: Yes, a good point, no default assignment required here. I'll switch to `data_attributes = nil`.

File: app/assets/stylesheets/govuk_publishing_components/components/_cards.scss [hunk 16]
```diff
@@ -0,0 +1,89 @@
+@import "mixins/grid-helper";
+@import "mixins/prefixed-transform";
+
+.gem-c-cards__heading {
+  margin: 0 0 govuk-spacing(6) 0;
+}
+
+.gem-c-cards__list {
+  @include columns($items: 16, $columns: 1, $selector: "li");
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/assets/stylesheets/govuk_publishing_components/components/_cards.scss:20]: This relates to the issue I raised with our team this morning about the number of items being fixed to support IE11 which Ian has since mentioned [here](https://github.com/alphagov/govuk_publishing_components/pull/2624#pullrequestreview-889623142). 

I'll have a think about the best way forward.

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_cards.html.erb:13]: I think your aim is laudable but the same thing could be achieved by writing this guidance in the documentation for this component. Having a heading inside this component adds unnecessary complexity and as far as I can tell it's exactly the same as putting a regular heading component before it.

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 11 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_cards.html.erb:4]: Ah yes I see, the shared_helper returns a string like `h2` rather than an integer. This might be too complicated for this initial version of the component then. A future improvement might be to modify the shared_helper something like this:

```
def get_heading_level
  return [*1..6].include?(@heading_level) ? "h#{@heading_level}" : "h2" unless @heading_level.zero?

  "span"
end

def get_subheading_number
  @heading_level + 1 unless get_heading_level == "span"
end
```

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 2 above)
Comment (andysellick) [reply app/views/govuk_publishing_components/components/docs/cards.yml:129]: Yes, some of our code doesn't follow our current best practice unfortunately. I'll make a note to find some time to clean up this and any other examples I can find, thanks for finding that one.

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 12 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_cards.html.erb:43]: Sorry my mistake, I see now that it's the closing bracket for line 17. Can I suggest you separate the `<%` and `%>` from the other code to make this clearer? e.g.

```
<% 
   items.each do |item|
      ...

   link = capture do
       link_to(link[:text], link[:path], {
         class: "govuk-link gem-c-cards__link",
         data: data_attributes,
       })
    end 
%>
```

File: spec/components/cards_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply spec/components/cards_spec.rb:76]: I don't think the tests need to be that granular. You could do a single test for 'all expected elements are present' and it would be clear enough that the one test was checking for multiple things, and it would be obvious where it had failed if it failed. Test files tend to get pretty long so I think making them smaller if possible is a good idea (I've had to completely rewrite some long test files and it can be hard going).

Something you might not be aware of - there's a [general test](https://github.com/alphagov/govuk_publishing_components/blob/master/spec/components/all_components_spec.rb#L26) that checks to see that every component template contains a class of the form `gem-c-[something]`, so you don't need to explicitly write a test to check for that.

**Comment by injms** [comment]
Other approaches that we could use:

 - drop the IE compatible grid and provide an alternative layout, as it's the IE variant of CSS Grid that's inflexible; this would mean that the grid is now flexible and reusable - but would mean a different layout for IE.
 - calculate the position of each element in the grid (in a [similar manner to the grid mixin][1]) in the template and then add that as inline CSS in a `style` element. If the stylesheet fails to load, the inline CSS will be ignored as the parent element won't be set to `display: grid`. This avoids sending every permutation of the grid layout to every user, regardless of whether it's needed
 - put the IE specific grid layout styles in an IE specific stylesheet, so only the browsers that need it will download the extra CSS.

[1]: https://github.com/alphagov/govuk_publishing_components/blob/a7845e9169fb334abfd3dcda9629aa01049307e2/app/assets/stylesheets/govuk_publishing_components/components/mixins/_grid-helper.scss

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (alex-ju) [reply app/views/govuk_publishing_components/components/_cards.html.erb:13]: Considering these are navigational lists I find it reassuring to know that the component promotes the use of a landmark as a way of reaching out to this section. [contents list](https://components.publishing.service.gov.uk/component-guide/contents_list), and [summary list](https://components.publishing.service.gov.uk/component-guide/summary_list) are promoting the use of headings in a similar way.

File: spec/components/cards_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (alex-ju) [reply spec/components/cards_spec.rb:43]: I can see how the level of test granularity if very different for this repo compared to `govuk-frontend` – maybe a point of discussion for one of the following community meetings, as we currently don't have any specific guidance on what good looks like for an rspec test (at least none that I'm aware of).

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (alex-ju) [reply app/views/govuk_publishing_components/components/docs/cards.yml:28]: My impression was that we favour an accessible example over a minimal one – this mental model made us [update the default radio example](https://github.com/alphagov/govuk_publishing_components/pull/2485). Happy to discuss this within the community though.

[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/docs/cards.yml:28]: There's a separate thread about the need for the heading. I'd hope that every example was accessible though...

**Comment by hannalaakso** [comment]
I've made the grid more flexible by setting the grid properties in the component (instead of using the `columns` mixin which polyfills the grid for IE11) and using `grid-template-rows: auto;` which calculates the number of rows required. The component now renders a single column on desktop for browsers that don't support css grid (primarily IE11) but constraints to 50% width, as recommended by our designer, to improve usability (especially for users of screen magnifiers).

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+<% if items.present? %>
+  <div class="gem-c-cards">
+    <% if heading %>
+      <%= content_tag(shared_helper.get_heading_level, class: "gem-c-cards__heading govuk-heading-m") do %>
+        <%= heading %>
+      <% end %>
+    <% end %>
+    <ul class="gem-c-cards__list<% if two_column_layout %> gem-c-cards__list--two-column-desktop<% end %>">
+      <% items.each do |item|
+
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:19]: I think if our linter doesn't enforce it I'm tempted to leave it as another developer could just change it again. Maybe discussing any rules our linter doesn't yet enforce but should do would be another good one for the community meeting.

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:13]: Since there is precedence for including a heading for a number of similar components and it promotes best practise and at least prompts teams to think about including a heading, I'm going to mark this  conversation as resolved. The component is still experimental and we can always revisit this if we get feedback to the contrary.

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml:22]: Since there is precedence in this repo for including a heading or other best practise elements in the first example I'm going to resolve this conversation, similar to https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r816626444. The component is still experimental and we can always revisit this if we get feedback to the contrary for what we do here.

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 1 above)
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml]: I've tweaked the wording to explain that including a heading is best practise.

File: spec/components/cards_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb:12]: Following our community discussion, we agree that there is a currently undocumented convention with GOV.UK Publishing components to only test markup where the template contains logic (eg. conditionals or forks), rather than testing for the integrity of the generated markup eg. that the correct classes and elements are generated even where there is no explicit template logic involved.

We do have visual regression tests and do PR reviews so those give us additional confidence.

However we might wish to consider testing for the integrity of the markup in the future, especially if we saw instances where this could have flagged unwanted changes. I've removed some tests to align with our current convention but left their removal out as a separate commit in case we want to revisit this later.

File: spec/components/cards_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb]: See https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r816938825

File: spec/components/cards_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb]: See https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r816938825.

File: spec/components/cards_spec.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (hannalaakso) [reply spec/components/cards_spec.rb]: See https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r816938825.

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 2 above)
Comment (andysellick) [reply app/views/govuk_publishing_components/components/docs/cards.yml]: I see you've renamed this but it still explicitly mentions tracking, which I'd like to avoid. For preference I think examples like this (and the code it refers to) should use the terminology of `data_attributes` rather than `tracking_attributes`. 

I appreciate that it's a pain to rename it at this point and there may be other components that don't follow this convention but I'm going to be looking into this soon to try to sort out any strays from this convention, and changing this now before it is put into use is going to be much easier than changing it later.

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 18]
```diff
@@ -0,0 +1,61 @@
+<%
+  heading ||= nil
+  items ||= nil
+  sub_heading_level = sub_heading_level || 3
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_cards.html.erb:4]: Nitpick: could this be written as `sub_heading_level ||= 3`?

**Review by andysellick — COMMENTED** [review]
Just a few minor things. I've re-commented on some of my earlier comments as well, but definitely looking good now 👍

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+        if !link[:path].present?
+          raise ArgumentError, "The cards component requires a href for all the links"
+        end
+
+        data_attributes = nil
+        attributes = link[:tracking_attributes] || nil
+
+        if attributes && attributes[:track_category] && attributes[:track_action]
+          data_attributes = {
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:30]: You're right, this is currently quite tightly coupled to the known instances where we'll be using tracking attributes (see https://github.com/alphagov/govuk_publishing_components/pull/2624/commits/5bf4e1544b808ae446d099c9df6de6f8b4423679). The other components like the select and image card seem to be more loosely defined and allow for any attributes to be passed through.

As we'd like to unblock this work and can revisit this on a future date, would it make sense for me to raise an issue about this so that we could get the PR merged?

File: app/views/govuk_publishing_components/components/docs/cards.yml (hunk 2 above)
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/docs/cards.yml]: We could look at this as part of the card for the discussion here: https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r817495300

I've also just added a description to this example to explain that no tracking events fire within the component guide.

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 19 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_cards.html.erb:29]: Yes that seems reasonable.

[the remark this answers is not in the document]
Comment (hannalaakso) [reply app/views/govuk_publishing_components/components/_cards.html.erb:29]: Card here: https://github.com/alphagov/govuk_publishing_components/issues/2656

File: app/assets/stylesheets/govuk_publishing_components/components/_cards.scss [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+    @supports (display: grid) {
+      width: initial;
+    }
+  }
+}
+
+.gem-c-cards__list--two-column-desktop {
+  @include govuk-media-query($from: "tablet") {
+    grid-template-columns: fractions(2);
+  }
```
Comment (injms) [inline app/assets/stylesheets/govuk_publishing_components/components/_cards.scss:45]: Isn't this the same as lines 22 to 24?

https://github.com/alphagov/govuk_publishing_components/blob/72fcb930d1a78c51ce33f8e1bab23a7e8c712592/app/assets/stylesheets/govuk_publishing_components/components/_cards.scss#L22-L24

If this needs to be two columns on desktop, should it be changes to `$from: "desktop"`?

```suggestion
  @include govuk-media-query($from: "desktop") {
    grid-template-columns: fractions(2);
  }
```

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+  shared_helper = GovukPublishingComponents::Presenters::SharedHelper.new(local_assigns)
+%>
+<% if items.present? %>
+  <div class="gem-c-cards">
+    <% if heading %>
+      <%= content_tag(shared_helper.get_heading_level, class: "gem-c-cards__heading govuk-heading-m") do %>
+        <%= heading %>
+      <% end %>
+    <% end %>
+    <ul class="gem-c-cards__list<% if two_column_layout %> gem-c-cards__list--two-column-desktop<% end %>">
```
[the remark this answers is not in the document]
Comment (injms) [reply app/views/govuk_publishing_components/components/_cards.html.erb:16]: +1 on this suggestion - other components try to keep the logic at the top of the partial template where possible, so it'd be good to keep that convention going in this component.

File: app/views/govuk_publishing_components/components/_cards.html.erb [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+      <%
+        items.each do |item|
+        link = item[:link]
+
+        if !link[:path].present?
+          raise ArgumentError, "The cards component requires a href for all the links"
+        end
+
+        data_attributes = nil
+        attributes = link[:tracking_attributes] || nil
```
Comment (injms) [inline app/views/govuk_publishing_components/components/_cards.html.erb:26]: Non blocking suggestion - Rails has a nicer way of doing this with `presence`:

```suggestion
        attributes = link[:tracking_attributes].presence
```

File: app/views/govuk_publishing_components/components/_cards.html.erb (hunk 21 above)
[the remark this answers is not in the document]
Comment (chao-xian) [reply app/views/govuk_publishing_components/components/_cards.html.erb:16]: +1

File: app/views/govuk_publishing_components/components/docs/cards.yml [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+  -  use the correct heading level hierarchy eg. if the component heading uses a `<h2>`, the subheadings must use a `<h3>`
+
+  It is also good practise to include a heading above the list to tell users what the list contains
+shared_accessibility_criteria:
+  - link
+examples:
+  default:
+    data:
+      heading: Topics
+      items:
```
Comment (injms) [inline app/views/govuk_publishing_components/components/docs/cards.yml:23]: A useful YAML thing is to name a section of the data: 
```yaml
data: &default-example-data
  items:
   - link:
        text: Benefits
        path: http://www.gov.uk
      description: Includes eligibility, appeals, tax credits and Universal Credit
```

and then you can reference that everywhere you want to use the `items` array:

```yaml
data:
  two_column_layout: true
  <<: *default-example-data
```

A real life example is the [link component docs][1].

[1]:https://github.com/alphagov/govuk_publishing_components/blob/019cd86bc63c985703c796ecb1b75133853d70a8/app/views/govuk_publishing_components/components/docs/list.yml

**Review by injms — APPROVED** [review]

**Review by injms — APPROVED** [review]
Dismissed by previous review, as this now is a much more flexible component - thanks to the use of `grid-template-rows: auto;`[^1] the component can now accept any number of items, which was my main concern.

I think the dropping of the inflexible version of the grid supported by IE10 and IE11 and switch to a fallback layout is a good decision - we've all now got to be careful when this component is updated that the fallback layout is thoroughly checked.

Super work @hannalaakso! Congratulations on your first component in the gem 🎉 

[^1]: https://github.com/alphagov/govuk_publishing_components/blob/72fcb930d1a78c51ce33f8e1bab23a7e8c712592/app/assets/stylesheets/govuk_publishing_components/components/_cards.scss#L17

**Review by andysellick — APPROVED** [review]
